flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -911,9 +915,11 @@ void releaseResources(boolean canceled) {
 			if (canceled) {
 				List<StateObject> statesToDiscard = new ArrayList<>();
 
-				statesToDiscard.add(metaStateHandle);
-				statesToDiscard.addAll(miscFiles.values());
-				statesToDiscard.addAll(newSstFiles.values());
+				synchronized (this) {
```

Comment:
Having a second look at this, the suggestion from the jira is not going far enough. 

We actually must ensure that all checkpointing efforts (e.g. writing to HDFS) have already come to an end, so that no new state handles can be created anymore at this point. Otherwise, even if we synchronize, the parallel thread could still create a new state handle after the getting back the lock.

However, once this is ensured, also the need for synchronization goes away. So I suggest that we rather make sure that cleanup is only performed after all checkpointing ended, and that no more checkpointing through this object can run after the cleanup.